src/Sparrow/Json/BlittableJsonReaderObject.cs

```diff
@@ -23,18 +23,77 @@ public sealed unsafe class BlittableJsonReaderObject : BlittableJsonReaderBase,
         private byte* _metadataPtr;
         private readonly int _size;
         private readonly int _propCount;
-        private readonly long _currentOffsetSize;
-        private readonly long _currentPropertyIdSize;
+
         private readonly bool _isRoot;
         private readonly byte* _objStart;
         private readonly byte* _propNames;
+
+        private readonly int _currentOffsetSize;
+        private readonly int _currentPropertyIdSize;
         private readonly int _propNamesDataOffsetSize;
 
+        private readonly int _functionTableIndex;
+
         public DynamicJsonValue Modifications;
 
         private Dictionary<StringSegment, object> _objectsPathCache;
         private Dictionary<int, object> _objectsPathCacheByIndex;
 
+
+
+
+        static BlittableJsonReaderObject()
+        {
+            // PERF: We will create a function table that ensures that it has enough space to handle the mapping. 
+            FunctionTable = new FunctionTableItem[0b01_000000];
```

Comment:
Can you please elaborate more about that? What is the motivation? What we want to avoid / achieve?

Reply:
Yes, what you are seeing here is the aftermath of several rounds of improvements and different ways to achieve the same thing. This is quite a neat hack which you will see being used at the Voron level extensively in the near future (I have been working on making this approach manageable for a long time).

If you follow the sequence of 4 commits related to this starting at https://github.com/ravendb/ravendb/pull/17411/commits/1e261d977c50f6d243b5ac25255905a286f237fe
You will see a pretty important thing, most of the behavior is already defined by the time we read the object from the storage. So most of the branches that we need to execute for example in `ReadNumber` are already defined on construction. 

The first iteration looked to avoid checking at runtime. What we do is package big operations that use that knowledge under a pattern that force the JIT to code generate for that specific behavior. We check that at the public call level and then do a dispatch to the specialized code.   For example: https://github.com/ravendb/ravendb/pull/17411/commits/1e261d977c50f6d243b5ac25255905a286f237fe#diff-e369f221bd22ce32476af1f71eedf88359da21d110397be3e186bd67f70e6156R226

The next improvement was acquiring the actual function pointers on construction (when we set those values) in such a way that now every object has it's own dynamic dispatch. When a public call is executed then that public call gets inline because the only thing that it does is call the function pointer the object has as a field. The cost of acquiring those function pointers for this high traffic object constructor could be improved. 

The last improvement was realize that we could just preallocate those options and store them on a function table, and just store at the object level which bucket of the function table we need to address to do it. We essentially implemented the dynamic dispatch table used in C++ inheritance.